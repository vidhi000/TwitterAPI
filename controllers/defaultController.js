const DefaultError = (ctx)=>{
    ctx.body = {
        msg : "Page Not Found 404"
      }
      ctx.status = 404

}

export {DefaultError}