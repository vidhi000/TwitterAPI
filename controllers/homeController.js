const Home = (ctx)=>{
    ctx.body = {msg : "This is Home Page"}
}

export {Home}