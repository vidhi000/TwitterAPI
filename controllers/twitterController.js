import dotenv from "dotenv";
import axios from "axios";
dotenv.config();

const Bearer = process.env.Bearer;
// let url = '/hashtag/:tag'
let valTag = /^[^ !@#$%^&*(),.?":{}|<>]{2,20}$/;

export const hashtagSearch = async (ctx) => {
  let reqTAG = ctx.request.URL.pathname.split("/");
  // console.log(reqTAG);
  const requestTAG = reqTAG[2];
  // console.log(requestTAG);

  if (!requestTAG.match(valTag)) {
    const err = {
      message: "HashTag is incorrect",
      success: false,
    };
    ctx.body = err;
  } else if (requestTAG.match(valTag)) {
    const hashtag = ctx.params.tag;
    const config = {
      headers: {
        Authorization: `Bearer ${Bearer}`,
      },
    };

    const ans = await axios.get(
      `https://api.twitter.com/2/tweets/search/recent?query=%23${hashtag}&max_results=15&tweet.fields=created_at&expansions=author_id,attachments.media_keys&media.fields=duration_ms,height,media_key,preview_image_url,public_metrics,type,url,width,alt_text`,
      config
    );

    const data = ans.data;
    // console.log(data);

    if (data.meta.result_count == 0) {
      ctx.body = {
        message: "Data is Not Found",
      };
    } else {
      let tweetArray = []
      for(let i = 0;i<5;i++){
        const username = data.includes.users[i].username;
      console.log(username);
      const { text: caption, created_at: date } = data.data[i];
      const details = { username, date, caption };

      let media_URL = "";
      let URL = "";
      if (data.data[i].attachments) {
        media_URL = data.data[i].attachments.media_keys[i];
        // console.log(media_URL);
        data.includes.media.forEach((ele) => {
          if (ele.url && ele.media_key === media_URL) {
            URL = ele.url.split(",");
          } else if (ele.preview_image_url && ele.media_key === media_URL) {
            URL = ele.preview_image_url.split(",");
          }
        });
      }
      URL.length > 0 ? (details["url"] = URL) : null;
        tweetArray.push(details)
      }
      ctx.body = { success: true, tweetArray };
      
    }
  }
};
//  export {ans} 

