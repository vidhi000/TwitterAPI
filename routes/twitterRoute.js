import koa from "koa"
import Router from "koa-router"
import { DefaultError } from "../controllers/defaultController.js"
import { Home } from "../controllers/homeController.js"
import { hashtagSearch } from "../controllers/twitterController.js"

const router = new Router()

router.get("/Home",Home)
router.get('/hashtag/:tag',hashtagSearch)
router.all(/.*/,DefaultError)

export {router} 