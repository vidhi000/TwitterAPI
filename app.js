import Koa from "koa";
import koaBodyparser from "koa-bodyparser";
import dotenv from "dotenv";
import {router} from "./routes/twitterRoute.js";

const app = new Koa();

dotenv.config();
const PORT = process.env.PORT

// app.use(koaBodyparser());

app.use(router.routes())

app.listen(PORT, () => {
  console.log(`server is runing on ${PORT}`);
});

export {app}

