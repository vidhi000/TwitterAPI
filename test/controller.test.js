import request from "supertest";
import { router } from "../routes/twitterRoute";
import koa from "koa";
import { hashtagSearch } from "../controllers/twitterController";
import app from "../app";
import expect from "expect"; 

describe("Testing", () => {
  let server;
  beforeAll(() => {
    const app = new koa();
    app.use(router.routes());
    app.use(router.allowedMethods());
    server = app.listen();
  });
  afterAll(() => {
    server.close();
  });

  test("Twitter API Testing", async () => {
    let response = await request(server)
      .get("/hashtag/india")
      .set("Accept", "application/json");
    // .expect('Content-Type', /json/)
    expect(response.status).toBe(200);
  });

  test("Home Controller Testing", async () => {
    let response = await request(server)
      .get("/Home")
      .set("Accept", "application/json");
    expect(response.status).toBe(200);
    expect(response.body.msg).toBe("This is Home Page");
  });


   test("Default",async()=>{
    let response = await request(server).get("/scscdc").set("Accept", "application/json");
    expect(response.status).toBe(404)
     expect(response.body).toMatchSnapshot()
    // console.log(response);
   })

   test("Incorrect Hashtag",async()=>{
    let response = await request(server).get("/hashtag/india%123").set("Accept", "application/json")
    expect(response.body).toMatchSnapshot()

   })

});
