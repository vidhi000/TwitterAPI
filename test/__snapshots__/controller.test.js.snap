// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Default 1`] = `
{
  "msg": "Page Not Found 404",
}
`;

exports[`Testing Incorrect Hashtag 1`] = `
{
  "message": "HashTag is incorrect",
  "success": false,
}
`;
